### PR TITLE
Handle pub(restricted)

### LINKED ISSUE #970: Handle new visibility types
syntex_syntax 0.31 brings a change from

```
pub enum Visibility {
    Public,
    Inherited,
}
```

to

```
pub enum Visibility {
    Public,
    Crate,
    Restricted { path: P<Path>, id: NodeId },
    Inherited,
}
```

It's unclear if the new variants need handling.

### PR DESCRIPTION
**Pull request by kamalmarhubi**
This commit properly handles pub(restricted) as introduced in RFC 1422
[0]. The syntax support was added in #971, but they were not correctly
formatted.

[0] https://github.com/rust-lang/rfcs/blob/master/text/1422-pub-restricted.md

Fixes #970

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -1817,7 +1814,7 @@ fn rewrite_where_clause(context: &RewriteContext,
 }
 
 fn format_header(item_name: &str, ident: ast::Ident, vis: &ast::Visibility) -> Option<String> {
```
Comment (marcusklaas) [inline src/items.rs]: This can no longer fail, right? I guess it should return a `String` then.

File: src/utils.rs [hunk 2]
```diff
@@ -66,14 +69,24 @@ pub fn extra_offset(text: &str, offset: Indent) -> usize {
     }
 }
 
+// Uses Cow to avoid allocating in the common cases.
 #[inline]
```
Comment (marcusklaas) [inline src/utils.rs]: This function isn't that trivial any more. Probably should leave it to the compiler to decide on inlining.

**Comment by marcusklaas** [comment]
This is a great pull request. Solid code, good test coverage and a nice application of `Cow`! I think we could use itertools' `join` in a few more spots, so the inclusion of that crate is useful too. I'll create an issue for that.

r+ with the two nits addressed

Thanks!

File: src/items.rs (hunk 1 above)
Comment (kamalmarhubi) [inline src/items.rs]: Done, and propagated up one more call to `format_unit_struct` :-)

File: src/utils.rs (hunk 2 above)
Comment (kamalmarhubi) [inline src/utils.rs]: Done!

**Comment by kamalmarhubi** [comment]
Thanks for the kind words! Re itertools, I am bringing it in in #1007 so I wasn't too bothered about using it here. Also they're all generic methods, so they won't increase code size except when used. It's a super handy crate!

**Pull request merged**